### ISSUE TITLE
Order again not firing `woocommerce_add_to_cart` hook when adding items to the cart.

### ISSUE THREAD

**Issue by claudiulodro** [woocommerce/woocommerce#21845]
**Describe the bug**
[This change to the order again handling](https://github.com/woocommerce/woocommerce/commit/759b99ad143bbe2a8cc7d156829dd5628ddfa616#diff-7bafcb4ec4e2600c4c4c5d2145ad8f8eR317) directly adds items to the cart array without ever firing the `woocommerce_add_to_cart` hook. This is causing issues on extensions/integrations that rely on that hook firing when items are added to the cart.

**Comment by helgatheviking** [comment]
@franticpsyx 👀

**Comment by franticpsyx** [comment]
We've suffered from this issue as well, and are about to roll out a patch.

Let's take a step back to think things through here :)

`woocommerce_add_to_cart` is designed to be fired after cart session data is loaded. Our "old" code used to add items to `WC()->cart->cart_contents` on `woocommerce_add_to_cart`. Firing that action before cart session data has been loaded would only trigger other/more issues. There's no real fix here because all existing code that hooks into this action expects it to run after session data has been loaded. 

Same with `woocommerce_add_cart_item` [here](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-cart-session.php#L330) -- this is likely another cause for alarm. I'd probably use a different filter name there.

Order-again is 3.5's biggest breaking change IMO: The context in which the old logic runs was moved before cart session load.

We've been wondering about the motivation behind this with @helgatheviking but weren't able to find the answer here -- maybe didn't look hard enough?

I see two solutions here:

1. Move the entire order-again logic _after_ session data has been loaded.
2. Treat this as a breaking change, write a blog post about it and avoid triggering hooks that run in `WC_Cart::add_to_cart` context there.

I'd personally opt for (1) because I know WCS also triggers order-again hooks _after_ session data has been loaded.

cc @thenbrent for feedback as this any code that hooks into WCS is affected. PB/CP/MnM are currently broken, FWIW.

**Comment by rrennick** [comment]
> for feedback as this any code that hooks into WCS is affected.

I checked the WCS support desk for issues reported in the last week that might be related to this change to `add_to_cart`  & did not find any.

**Comment by helgatheviking** [comment]
Could someone explain _why_ the logic was moved? That would help me evaluate the possible solutions here. Since I don't know I am just naively thinking it would be best to restore and undeprecate the `order_again()` form handler method.

**Comment by franticpsyx** [comment]
> I checked the WCS support desk for issues reported in the last week that might be related to this change to add_to_cart & did not find any

It may take a while to see the consequences, I think. WCS code is still triggering the `woocommerce_order_again_cart_item_data` and `woocommerce_add_to_cart_validation` filters in a post-session-load context, so all the existing implementations out there just work.

Once people start writing code that assumes these hooks are triggered before session load, things will probably change.

> Could someone explain why the logic was moved?

Same here, I can't figure out what problem it solves. Moving the logic out of the form-handler class was a solid idea, but it might have been safer to add the logic towards the end of `get_cart_from_session`, or in a dedicated method still hooked onto `wp_loaded`.

**Comment by rrennick** [comment]
> WCS code is still triggering the woocommerce_order_again_cart_item_data and woocommerce_add_to_cart_validation filters 

Thanks.  Agreed, they may take a while to show up.

**Comment by franticpsyx** [comment]
> Could someone explain why the logic was moved?

Looking at the code again I think it was a performance-conscious decision.

Previously, order-again logic was calling `WC_Cart::add_to_cart` for every order item, and therefore totals were re-calculated multiple times, and session data set multiple times as well.

The new approach makes things a lot faster, and I'm leaning towards considering this as a welcome change, despite the backwards-compatibility issues we're discussing.

I still think that `WC_Cart::add_to_cart` actions/filters have no place here:

* Session data is not loaded yet, so any code hooking into `woocommerce_add_cart_item` or `woocommerce_add_to_cart` is likely to trigger errors if it attempts to work with cart contents.
* `WC_Cart::add_to_cart` is designed to catch exceptions in order to allow third-party code to generate failure notices from code/callbacks hooked into those filters/actions.

So, summing up, I'd propose the following here:

1. Don't trigger any hooks that run in `WC_Cart::add_to_cart` context here.
2. Rename `woocommerce_add_cart_item` [here](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-cart-session.php#L329) to `woocommerce_add_order_again_cart_item` or remove the filter altogether.
3. Pass the `$cart` variable into the `woocommerce_ordered_again` action, as it's no longer possible to access anything through `WC_Cart::get_cart`.
4. Possibly write a short post on the Dev blog explaining the benefits of this change (💯) and the catch-22's.

**Comment by helgatheviking** [comment]
We could also add a duplicate trigger that mimics `woocommerce_add_to_cart` such as:

 `do_action( 'woocommerce_add_to_cart_order_again', $cart_id, $product_id, $quantity, $variation_id, $variations, $cart_item_data );`

at the end of the each loop, [here](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-cart-session.php#L342). 

Or, and this is more complex, refactor the `$cart->add_to_cart()` to include a `$context` of sorts that controls whether or not the totals need to be recalculated.

**Comment by claudiulodro** [comment]
Hi folks,

Thanks for further investigating this issue and documenting your discussion here. :)

The original aim was to begin refactoring/phasing out parts of the giant form handler class into a better system beginning with the "Order again" handling. Since that is used the least of all the actions in the form handler class, we can identify whether we'll run into big issues when we refactor the other actions without the mass chaos of accidentally making breaking changes to`add_to_cart_action`. This also provides a chance to improve performance and other such improvements to the handling of the actions.

My apologies for not documenting it better; I didn't realize how breaking the change was to certain extensions. We'll be bringing back dev chats soon, so that should give us a way to run potentially breaking changes by the developer community.

I agree with the consensus I'm reading here: firing the `woocommerce_add_to_cart` hook before the session is loaded will likely cause big problems, since extensions will behave as if the session had loaded. 

@franticpsyx's suggested changes sound like a good solution, and I'd be happy to review a PR if you want to give it a shot. The reports here are the only ones we've reveived about this change breaking extensions, so if the proposed solution works well for you, there should be no further issues 🤞.

Thanks!

**Comment by claudiulodro** [comment]
#21947

**Comment by helgatheviking** [comment]
I believe we have more work to do here to make the order again process more faithfully mimic the old process. In Mix and Match, even with @franticpsyx 's solution and patch, I still have multiple cases that result in inaccurate carts (documented [here](https://github.com/woocommerce/woocommerce-mix-and-match-products/issues/234)). Some of those are related to #22099, but ultimately I still believe it will be best if Mix and Match can fully control when/if the child items for a MNM container get added to the cart.

**Comment by helgatheviking** [comment]
@claudiulodro Thanks for re-opening this. Please let me know if there's anything I can do to help move this along.

**Comment by mikejolley** [comment]
`woocommerce_ordered_again` should be suitable for that because it fires once all items are added to the cart array. Order again cannot fully mimic regular cart adds because of the performance issue that causes.

There is also `woocommerce_add_to_cart_validation` which can be used to prevent the container item being added to the cart if a part in your case is not able to be added. You'd just need to return false there and optionally add an error message via `wc_add_to_cart_message()`. That should cover your needs.

I'm looking at #22099 now to see if that is patchable. The main issue there is avoidingly a series of costly stock checks which again would cause a performance bottleneck for something that could be considered edge case. The user cannot actually purchase the items in the cart so it's just an inconvenience to them.

**Comment by helgatheviking** [comment]
@mikejolley thanks for looking in to this. Can you advise how I could use `woocommerce_ordered_again` for a mix and match container to add it's children products to the cart? The `$cart` array is not passed by reference and `WC()->cart` is not set up yet. How can I interact with and modify the cart contents on this hook?

>There is also` woocommerce_add_to_cart_validation` which can be used to prevent the container item being added to the cart if a part in your case is not able to be added. 

The validation on the container itself is fine. Well I think so, it's been a while since I've looked at this. But now that the parent is not responsible for adding it's children (because there's no action after the addition of each item like their is in `add_to_cart()`) it's the child products that aren't being properly validated. The children should only be added _if_ the parent is valid, and they should be added right after the parent is added so they stay together in a customer-friendly way in the cart.

But now that WC adds everything from the order line by line, the child products can be added to the cart when the parent container is not valid. So then the parent/child relationship can be lost, which is an integral part of how Mix and Match, Product Bundles, and Composite Products all work.

**Comment by mikejolley** [comment]
> Can you advise how I could use woocommerce_ordered_again for a mix and match container to add it's children products to the cart?

Shouldn't those items already be in the order if they are being reordered?

**Comment by helgatheviking** [comment]
>Shouldn't those items already be in the order if they are being reordered?

Sort of. Here's an example:

Widget Container (any 4 widgets)
Red Widgets x 2
Blue Widgets x 2

If Blue Widgets are out of stock when reordering, then the Widget Container is not valid, and none of the items should be added to the cart. If the Widget Container is not valid it's not added, but when validating Red Widget there's currently no way to check if it's parent is in the cart because there isn't a cart yet. 

@franticpsyx has written something for Product Bundles and Composites that restores the parent/child connection, but I've [documented multiple cases](https://github.com/woocommerce/woocommerce-mix-and-match-products/issues/234) where it fails. 

What I'm trying to say is it makes our code a LOT simpler if the parent Mix and Match container can control when it needs to add it's own children... ie only when the entire container is valid.

**Comment by franticpsyx** [comment]
> For the container itself, you can check the stock levels of the children and return false on `woocommerce_add_to_cart_validation` as suggested.

+1

There are parent/child meta to establish these relationships, so this should be possible. @helgatheviking looking at this I don't think there's any way to get things to work as they did. Our "old" logic can't work before session data is loaded.

@mikejolley it might have helped to add the items straight into `cart_contents` instead of using a local array, because it would allow us to re-use existing code, I think. As it is, the only way to extend the logic is to use `woocommerce_add_to_cart_validation`, and log "rejections" on our end: There's no easy way for a child to see if its parent got added or not through `woocommerce_add_to_cart_validation` or any other filter there.

**Comment by mikejolley** [comment]
@franticpsyx that encourages folks to run logic every add, and rely on the order in which things are added. If the parent 'meta' is preserved when the item is loaded into session, could the check not come later (remove it if the parent did not populate)?

**Comment by franticpsyx** [comment]
@mikejolley yup, that's exactly what PB/CP are doing -- removing orphans on `woocommerce_cart_loaded_from_session` to account for such edge-case conditions.

Since PB/CP were patched, we haven't received any issues so without checking I assume that most of the issues Kathy found in MnM are handled by these sanity checks in PB/CP. Will need to confirm.

@helgatheviking if the parent gets rejected by `woocommerce_add_to_cart_validation `, would a cleanup on `woocommerce_cart_loaded_from_session` help with the issues you found?

**Comment by helgatheviking** [comment]
Could we add a filter to manipulate the order again `$cart` contents? ie:

```
$cart = apply_filters( 'woocommerce_ordered_again_cart_contents', $cart );
````

**Comment by helgatheviking** [comment]
Or replace the `do_action( 'woocommerce_ordered_again' )` with 

```
do_action_ref_array( 'woocommerce_ordered_again', array( $order->get_id(), $order_items, &$cart ) );
```
